### PR TITLE
Pins development

### PR DESCRIPTION
**Pull request by CorentinGaut**
Got a stable demo with pins object in the ud-vz scene : 
- Click on it to show details of the episode
- Add an episode class with specific CSS to build easily
- Compass

Next step : 
- Lock episode with a config file wich it will be specify to the progress of the user
- Search bar for episode 
- Tactile inputs

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -7,5 +7,6 @@
   </head>
   <body>
     <script src="./dist/release/app_name.js"></script>
+    </script>
```
Comment (EricBoix) [inline index.html]: Redundant script close

File: src/EpisodeVisualizer.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+export class EpisodeVisualizer { 
+    constructor(name, view3D = new udviz.Views.View3D()) {
+        this.name = name;
+        this.view3D = view3D;
+
+         this.pinsObject = new THREE.Sprite();    
+         //let windows = new udviz.Widgets.DocumentVisualizerWindow();  
+    }
+
+    //Create Pins
```
Comment (EricBoix) [inline src/EpisodeVisualizer.js]: NIce comment ;-P

File: src/Help.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
+/** @format */
+
+//Components
+import { ModuleView } from '../node_modules/ud-viz/src/Widgets/Components/ModuleView/ModuleView';
```
Comment (EricBoix) [inline src/Help.js]: Don't import with installation relative path but with something similar to
[`import { Widgets, Components } from 'ud-viz';`](https://github.com/VCityTeam/UD-Demo-DatAgora-Vegetalisation-PartDieu/blob/125e5b60e32efd111e729e40b1d3ad484f9e5c8b/ud-viz-context/BaseDemo.js#L4)

File: src/bootstrap.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    view3D.getScene(),
+    directionalLight,
+    ambientLight
+  );
+
+  //Help module
+  const help = new HelpWindow();
+
+  const center = view3D.getExtent().center();
+  const positionPins_1 = new udviz.THREE.Vector3(1843554.77, 5165405.73, 220);
```
Comment (EricBoix) [inline src/bootstrap.js]: Avoid hardcoding data in code but use a config file.

File: src/bootstrap.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  const positionPins_2 = new udviz.THREE.Vector3(1844804.07, 5168372.18, 260); 
+  const positionPins_3 = new udviz.THREE.Vector3(1843470.01, 5164312.59, 220);
+
+  //Test d'un episode visualizer
+  const episode_1 = new EpisodeVisualizer('episode_1', view3D);  
+  episode_1.createPin(positionPins_1,"../assets/img/Episode1_1.png","../assets/img/Episode1_1_lock.png",false);
```
Comment (EricBoix) [inline src/bootstrap.js]: In the above mentioned config file associate the pin position and the file (as opposed to separating them)

File: src/bootstrap.js [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+  //TO-DO make a list of object clickable
+  let listPins = episode_1.getPinsObject();
+
+  view3D.html().addEventListener( 'click', onDocumentMouseClick );
+  view3D.html().addEventListener( 'pointermove', onDocumentMouseLeave );
+
+  //Compass img element
+  const compass = document.createElement('img');
+  compass.src = '../assets/img/compass.png';
+  compass.style.position ='absolute';
```
Comment (EricBoix) [inline src/bootstrap.js]: Place this in a css file.

File: src/bootstrap.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+                                            
+    let raycaster =  new udviz.THREE.Raycaster();                                        
+    raycaster.setFromCamera( mouse3D, view3D.getCamera());
+
+    //All objects hits
+    let intersects = raycaster.intersectObjects( view3D.getScene().children ); 
+    if ( intersects.length > 0 ) {
+      if (!intersects[ 0 ].object.userData.LOCK){
+          divEpisode.style.setProperty('display','block');
+          //window.open("https://www.derrierelesfumees.com/_Contenusdlf/Episodes/Episodes01/index.html")
```
Comment (EricBoix) [inline src/bootstrap.js]: Remove this comment.

File: src/bootstrap.js [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    sph.setFromVector3(dir);
+    compass.style.transform = `rotate(${udviz.THREE.Math.radToDeg(sph.theta) - 180}deg)`;
+  });
+
+  /* --------------------------------- EVENT --------------------------------- */
+
+  //Show episode div
+  function onDocumentMouseClick( event ) {    
+    event.preventDefault();
+    let mouse3D = new udviz.THREE.Vector3( ( event.clientX / window.innerWidth ) * 2 - 1,
```
Comment (EricBoix) [inline src/bootstrap.js]: Use pickObjectsAt as [done here](https://github.com/VCityTeam/UD-Viz/blob/master/src/Widgets/Components/LayerManager/LayerManager.js#L135)

File: src/bootstrap.js [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+                            0.5 );     
+                                            
+    let raycaster =  new udviz.THREE.Raycaster();                                        
+    raycaster.setFromCamera( mouse3D, view3D.getCamera());
+
+    //All objects hits
+    let intersects = raycaster.intersectObjects( view3D.getScene().children ); 
+    if ( intersects.length > 0 ) {
+      if (!intersects[ 0 ].object.userData.LOCK){
+          intersects[ 0 ].object.material.color.set("rgb(200, 200, 200)");
```
Comment (EricBoix) [inline src/bootstrap.js]: Check the code of CityObjectProvider: on how they deal with color changes.

File: src/bootstrap.js [hunk 10]
```diff
@@ -0,0 +1,145 @@
+/** @format */
```
Comment (EricBoix) [inline src/bootstrap.js:1]: Think of promoting this code to ud-viz.[...]Widgets.

**Review by EricBoix — CHANGES_REQUESTED** [review]

File: src/bootstrap.js (hunk 5 above)
Comment (CorentinGaut) [reply src/bootstrap.js]: Done

File: src/bootstrap.js (hunk 6 above)
Comment (CorentinGaut) [reply src/bootstrap.js]: Done

File: src/bootstrap.js (hunk 8 above)
Comment (EricBoix) [reply src/bootstrap.js]: @CorentinGaut said: the created object does not seem to appear in the list of pickable object.

File: src/bootstrap.js (hunk 9 above)
Comment (EricBoix) [reply src/bootstrap.js]: Same issue as above.

File: src/bootstrap.js (hunk 8 above)
Comment (EricBoix) [reply src/bootstrap.js]: Issue #4 opened.

**Review by EricBoix — APPROVED** [review]

File: src/EpisodeVisualizer.js [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+ * Class to create an episode with all his content materialized in ud-viz scene
+ * 
+ */
+export class EpisodeVisualizer { 
+  /**
+   * Create an episode
+   * 
+   * @param {string} name name of your episode
+   * @param {View3D} view3D the view where you put all your content 
+   * @param {JSON} episodeConfig JSON data of your episode
```
Comment (EricBoix) [inline src/EpisodeVisualizer.js]: Promote this JSON to a native JS object. The usage made of the episodeConfig json
string is only to create an EpisodeContent. Pass this EpisodeContent as argument to
this constructor instead ?

**Review by EricBoix — APPROVED** [review]

File: README.md [hunk 12]
```diff
@@ -1,26 +1,144 @@
...
+  * Installing with the [CLI](https://en.wikipedia.org/wiki/Command-line_interface)
+
+    ```bash
+    iex (new-object net.webclient).downstring(‘https://get.scoop.sh’)
+    scoop install nodejs
+    ```
+
+## Installing and running the template application
+
+The template application can be locally (on your desktop) started in the following way
```
Comment (EricBoix) [inline README.md]: All what concerns the original ud-viz-template should be declined to this demo.

**Review by EricBoix — CHANGES_REQUESTED** [review]

**Review by EricBoix — APPROVED** [review]
Some work deported to newly opened issues e.g. #6

File: README.md (hunk 12 above)
Comment (EricBoix) [reply README.md]: Place in issue #7

**Pull request merged**